### ISSUE TITLE
nan issue when minimal_damping is small

### ISSUE THREAD

**Issue by tiancaiwyj** [google-deepmind/kfac-jax#183]
Hi, 

I am adapting the QuickStart example to solve some simple physics-informed neural network problem, which is basically a regression problem (using squared_error_loss) plus a physical equation added into the loss function.

My `l2_reg` is set to be 0.  The convergence rate of my training looks awesome when setting a small `minimal_damping `value. For example, when setting `min_damping = 1e-15`. My training loss can reach 1e-17 after 2000 iterations. However, if I further reduce the `min_damping`, then after certain iterations, the training loss and other parameters of the optimizer becomes `nan` (see below).

 I was wondering if it is because the calculation within the optimizer reaches machine precisions, and is there any way to further lower the `min_damping` value while not have the `nan` issue.

below is my test code:
```python
import jax.config as config
import haiku as hk
import jax
import jax.numpy as jnp
from jax import vjp
import kfac_jax
import matplotlib.pyplot as plt

config.update('jax_enable_x64', True)

# Hyper parameters
NUM_OUTPUT = 1
L2_REG = 0

def make_dataset_iterator(batch_size):
  # dataset for regression problem
  x = jnp.linspace(0, 1, batch_size)[:, None]
  y = jnp.exp(x)
  return x, y


def squared_error(pred, targets=None):
  """squared_error loss."""
  # Tell KFAC-JAX this model represents a classifier
  kfac_jax.register_squared_error_loss(pred, targets)
  loss = (pred-targets) ** 2
  return loss


def model_fn(x):
  """A Haiku MLP model function - three hidden layer network with tanh."""
  return hk.nets.MLP(
    output_sizes=(20, 20, 20, NUM_OUTPUT),
    with_bias=True,
    activation=jax.nn.tanh,
  )(x)


# The Haiku transformed model
hk_model = hk.without_apply_rng(hk.transform(model_fn))


# define the governing equation to solve
def gov_eqn(func, x):
    u, vjp_fn = vjp(func, x)  # Obtain its gradient function
    u_x = vjp_fn(jnp.ones(x.shape))[0]
    f = u_x - u
    return f, u


def loss_fn(model_params, model_batch):
  """The loss function to optimize."""
  x, y = model_batch
  predf = lambda x: hk_model.apply(model_params, x)
  f, u_p = gov_eqn(predf, x)

  # data loss at the boundary condition
  loss1 = jnp.mean(squared_error(u_p[0], y[0]))
  # equation residue
  loss2 = jnp.mean(squared_error(f, 0))

  return loss1 + 0.1 * loss2


# Create the optimizer
optimizer = kfac_jax.Optimizer(
  value_and_grad_func=jax.value_and_grad(loss_fn),
  l2_reg=L2_REG,
  value_func_has_aux=False,
  value_func_has_state=False,
  value_func_has_rng=False,
  use_adaptive_learning_rate=True,
  use_adaptive_momentum=True,
  use_adaptive_damping=True,
  initial_damping=1e-4,
  min_damping=1e-16,
  multi_device=False,
)

input_dataset = make_dataset_iterator(400)
rng = jax.random.PRNGKey(1234)
x_input, y_target = input_dataset
rng, key = jax.random.split(rng)
params = hk_model.init(key, x_input)
rng, key = jax.random.split(rng)
opt_state = optimizer.init(params, key, input_dataset)

# Training loop
for l in range(2500):
  rng, key = jax.random.split(rng)
  params, opt_state, stats = optimizer.step(
      params, opt_state, key, batch=input_dataset, global_step_int=l)
  print(l, stats)
```

and below is the `nan` issue that start at iteration 1655 when setting `min_damping = 1e-16`
```
1654 {'batch_size': Array(400, dtype=int32), 'damping': Array(1.e-16, dtype=float64), 'data_seen': Array(662000, dtype=int64), 'learning_rate': Array(1.13615166e-05, dtype=float64), 'loss': Array(4.79548902e-17, dtype=float64), 'momentum': Array(0.87433691, dtype=float64), 'new_loss': Array(4.78584102e-17, dtype=float64, weak_type=True), 'quad_model_change': Array(-5.09050911e-20, dtype=float64), 'rho': Array(1.89528981, dtype=float64, weak_type=True), 'scaled_grad_norm_sq': None, 'step': Array(1655, dtype=int64)}
1655 {'batch_size': Array(400, dtype=int32), 'damping': Array(1.e-16, dtype=float64), 'data_seen': Array(662400, dtype=int64), 'learning_rate': Array(nan, dtype=float64), 'loss': Array(4.78584101e-17, dtype=float64), 'momentum': Array(nan, dtype=float64), 'new_loss': Array(nan, dtype=float64, weak_type=True), 'quad_model_change': Array(nan, dtype=float64), 'rho': Array(nan, dtype=float64, weak_type=True), 'scaled_grad_norm_sq': None, 'step': Array(1656, dtype=int64)}
```
